lgc/patch/PatchResourceCollect.cpp

```diff
@@ -301,6 +328,13 @@ bool PatchResourceCollect::canUseNggCulling(Module *module) {
     return false;
   }
 
+  // Check resource usage, disable NGG culling if there are resource write operations (including atomic operations).
+  // on non-GS shaders. This is because such write operations have side effect in execution sequences.
+  const auto resUsage = m_pipelineState->getShaderResourceUsage(
+      hasGs ? ShaderStageGeometry : (hasTs ? ShaderStageTessEval : ShaderStageVertex));
+  if (!hasGs && resUsage->resourceWrite)
+    return false;
```

Comment:
I don't quite understand this logic. Do we really have to disable NGG here, or do we only have to disable *NGG culling* because culling a vertex makes its stores disappear?

Reply:
I revise the codes. The comments doesn't well explain the reason. Let me have some additional clarification.

For non-GS NGG, we have deferred attribute shading so the HW VS is partially executed before culling (get position export data) and partially executed after culling (get non-position export data). If there are write operations like memory store and atomic scattered in the shader, NGG culling should be disabled. Partial execution will have impacts on the execution sequences. This is the root cause why the CTS fails. In the future, when I finish reworking non-GS NGG, we can allow one case - shader just exporting position (no other export). 

Why GS NGG doesn't have such problem? GS NGG culling is done after ES/GS execution. At that time, all export values are written to GS-VS ring and culling is based on position data of the output vertices so GS NGG is immune to this defect.